preshed/bloom.pyx

```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    # legacy deserialization - otherwise length is always a multiple of
+    # KEY_BITS.
+    buflen = bloom.length // KEY_BITS
+    if bloom.length % KEY_BITS > 0:
+        buflen += 1
+    contents = [bloom.bitfield[i] for i in range(buflen)]
+    buffer = struct.pack(f"<{buflen}Q", *contents)
+    return prefix + buffer
 
 
 cdef void bloom_from_bytes(Pool mem, BloomStruct* bloom, bytes data):
-    py = array("L")
-    if hasattr(py, "frombytes"):
-        py.frombytes(data)
-    else:
-        py.fromstring(data)
-    bloom.hcount = py[0]
-    bloom.length = py[1]
-    bloom.seed = py[2]
-    bloom.bitfield = <key_t*>mem.alloc(bloom.length // sizeof(key_t), sizeof(key_t))
-    for i in range(bloom.length // sizeof(key_t)):
-        bloom.bitfield[i] = py[3+i]
+    # new-style memory structure (each unit is a key_t/64 bits):
+    # - pad1: 0 (not valid in old data)
+    # - ver: 1 (can be raised later if necessary
+    # (following values are same as old-style, except length semantics changed)
+    # - hcount: number of hashes
+    # - length: bitfield length in bits
+    # - seed: seed value for hashes
+
+    if len(data) < 40:
+        # unlikely but possible with old data
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    pad, ver, hcount, length, seed = struct.unpack("<QQQQQ", data[0:40])
+    if pad !=0:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    assert ver == 1, "Unknown serialization version"
+
+    bloom.hcount = hcount
+    bloom.length = length
+    # To avoid overflow, just take the low bits. In valid data nothing will be
+    # lost.
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
+    bloom.seed = safe_seed
+
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+    for i in range(buflen):
+        bloom.bitfield[i] = contents[i]
+
+
+cdef void bloom_from_bytes_legacy(Pool mem, BloomStruct* bloom, bytes data):
+    # Older versions of this library used the array module with type L for
+    # serialization. Types in array guarantee a minimum size, not an actual
+    # size, and it turns out L is 8 bytes on Linux and most platforms, but 4 on
+    # Windows. 
+
+    # As a separate issue, due to bits/bytes confusion, each container in the
+    # serialized data has only one byte actually used.
+
+    # The code in this function reads in data in the old format and converts it
+    # to the current format losslessly. It also packs the significant bytes into
+    # contiguous memory.
+
+    # on non-Windows platforms
+    unit = "Q"
+    unit_size = 8 # size of container in bytes
+    offset = unit_size * 3
+    hcount, length, seed = struct.unpack("<QQQ", data[0:offset])
+
+    decode_len = length // unit_size # number of units to unpack
```

Comment:
Isn't `decode_len`

```suggestion
    decode_len = length // KEY_BITS
```

in both cases?

Reply:
You are correct that it is  the same in both cases, but `length` is in terms of bytes, so `KEY_BITS` cannot be the denominator. (You can also see that using it will make the tests fail.)

Reply:
Let me rephrase that: the units of `length` are confusing due to the bug, but because `decode_length` is used in an `unpack` call later, it needs to be expressed in terms of the container type used for unpacking, which is drawn from the legacy data.